library/Stratosphere/Values.hs

```diff
@@ -22,38 +21,27 @@ import Data.String (IsString (..))
 import Data.Text (Text)
 import GHC.Exts (IsList(..))
 
--- GADTs are cool, but I couldn't get this to work with FromJSON. Now that we
--- don't have FromJSON though, we can revisit this.
-
--- data Val a where
---   Literal :: a -> Val a
---   Ref :: Text -> Val a
---   If :: Text -> Val a -> Val a -> Val a
---   And :: Val Bool -> Val Bool -> Val Bool
---   Equals :: (Show a, ToJSON a) => Val a -> Val a -> Val Bool
---   Or :: Val Bool -> Val Bool -> Val Bool
-
 -- | This type is a wrapper around any values in a template. A value can be a
 -- 'Literal', a 'Ref', or an intrinsic function. See:
 -- http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html
-data Val a
- = Literal a
- | Ref Text
- | If Text (Val a) (Val a)
- | And (Val Bool) (Val Bool)
- | Equals (Val a) (Val a)
- | Or (Val Bool) (Val Bool)
- | GetAtt Text Text
- | Base64 (Val Text)
- | Join Text (ValList Text)
- | Select Integer (ValList a)
- | FindInMap (Val a) (Val a) (Val a) -- ^ Map name, top level key, and second level key
- | ImportValue Text -- ^ The account-and-region-unique exported name of the value to import
- | Sub Text (Maybe (HashMap Text (Val Text))) -- ^ Substitution string and optional map of values
+data Val a where
+  Literal :: a -> Val a
+  Ref :: Text -> Val a
+  If :: Text -> Val a -> Val a -> Val a
+  And :: Val Bool' -> Val Bool' -> Val Bool'
+  Equals :: (Show a, ToJSON a) => Val a -> Val a -> Val Bool'
+  Or :: Val Bool' -> Val Bool' -> Val Bool'
+  GetAtt :: Text -> Text -> Val a
+  Base64 :: Val Text -> Val Text
+  Join :: Text -> ValList Text -> Val Text
+  Select :: Integer -> ValList a -> Val a
+  FindInMap :: Val a -> Val a -> Val a -> Val a -- ^ Map name, top level key, and second level key
```

Comment:
Not sure about this type. Looks like the first argument is `Val Text`, the second is some kind of key-value list (??), and the third is also (maybe) `Val Text`

Reply:
You're right, most of these should be `Val Text`. Fixed.

```haskell
FindInMap :: Val Text -> Val Text -> Val Text -> Val a -- ^ Map name, top level key, and second level key
```